Allow to manually reorder "normal" folders

pabzm opened the pull request:
Protected folders like INBOX, Sent, etc. are staying at the top of the list.

Fixes #5003

TODO:
- [x] Fix broken state after folder renaming
- [x] Fix the styling (equalize white space around list elements through padding, avoid ID as selector)
- [x] Figure out if the item was moved between protected folders, which is not allowed.
- [x] ~~Save only if the folder list is different than at start~~ (Not worth the effort)
- [ ] In the receive callback, can we wait for the confirmation dialog without introducing async/await and Promises?


Screenshots:

<img width="405" height="534" alt="Screenshot 2025-07-25 at 16 24 28" src="https://github.com/user-attachments/assets/791592df-dbb0-443e-98f3-2563eb78f0c6" />

---------

<img width="401" height="505" alt="Screenshot 2025-07-25 at 16 24 37" src="https://github.com/user-attachments/assets/07bbd353-35ac-41e6-a3aa-ba29224b12f5" />


---------

<img width="693" height="577" alt="Screenshot 2025-07-25 at 16 24 49" src="https://github.com/user-attachments/assets/5cc2db5a-05ae-4268-9088-cd760c9c65cf" />


---------


<img width="408" height="544" alt="Screenshot 2025-07-25 at 16 24 57" src="https://github.com/user-attachments/assets/ffdad971-dbcd-4e63-ba48-7bf5d91e07d0" />

pabzm committed:
Allow to manually reorder "normal" folders

Special folders like INBOX, Sent, etc. are staying at the top of the
list.

johndoh commented:
I like the idea of allowing users to reorder folders. I think having different drag behavior depending where on the <li> element you click is a bit clunky, it would be nice if both functions (change parent and change order) could be combined somehow. Putting the drag icon next to the subscription toggle could lead to fat finger problems, I suggest putting it at the start, something like this:

<img width="842" height="377" alt="Screenshot 2025-07-10 183123" src="https://github.com/user-attachments/assets/de276f91-e36a-42d8-8d20-e28d3e9d8c25" />

Out of scope for this PR but adding a similar drag icon the rule list in the managesieve plugin would also be nice, highlight to the users dragging is possible - so the implementation should be done is such a way the drag icon can be added to any list.

pabzm committed:
WIP: Allow to manually reorder "normal" folders using $.sortable()

pabzm commented:
I found a technical solution to combine both actions, reordering and moving into (and out of) subfolders. It will take some more working days to adapt it to Roundcube's code, though.

pabzm committed:
Make close_func being called if the dialog is closed, not cancelled

pabzm committed:
Use Promise to enable waiting for the decision from the dialog

pabzm committed:
Fix showing descendent elements beyond the borders of the li parent

Now a possible element overflow is hidden on the link element, not the
li element.
Without this, a child element (from a sub-list) would be invisible if
dragged outside of the li element.

pabzm committed:
Make moving a folder to the bottom less fiddly

pabzm committed:
Improve colors with hovering in light and dark mode

pabzm commented:
@johndoh Thank you for the feedback! I improved the styling in dark (and light) mode a little, and gave the list's a bit more space at their end to make it easier to drop elements there.

> When I drag a sub folder from one parent to another it saves but the UI does not update, have to refresh the page.

I cannot reproduce that. Could you check for errors in the browser console or anything else that appears remarkable when that happens?

(I'll continue work on this in ~2 weeks.)

pabzm commented:
@Neustradamus The default and fallback is alphabetical ordering as long as you don't start to order the folders manually.

johndoh commented:
I think may be the list not updating problem only happens when a custom order is first applied.

program/js/app.js:8183
`Uncaught ReferenceError: next_sibling is not defined`

for me this fixes the issue:
```patch
diff --git a/program/js/app.js b/program/js/app.js
index 08a57a0db..930741e4c 100644
--- a/program/js/app.js
+++ b/program/js/app.js
@@ -8163,11 +8163,9 @@ function rcube_webmail() {
             delete ref.env.subscriptionrows[fname];
         });

-        if (this.env.folder_ordered_manually) {
-            // We need to store this information now, because it's not available anymore after removing the row from
-            // the DOM.
-            next_sibling = row.nextElementSibling;
-        }
+        // We need to store this information now, because it's not available anymore after removing the row from
+        // the DOM.
+        next_sibling = this.env.folder_ordered_manually ? row.nextElementSibling : null;

         // get row off the list
         row = $(row).detach();
```

pabzm committed:
fixup! 255e51edef58e4881871d06eed70bde4fa122b9d

pabzm committed:
Prevent sortable folders to be sorted between protected folders

Remove outdated comment

pabzm committed:
Move folders up/down with buttons

pabzm committed:
Allow get_folder_li() for folder sorting

pabzm committed:
A little refactoring (folder_name2id())

pabzm committed:
Styling for sortable folders list

pabzm committed:
Fix code for PHP <8.4

pabzm committed:
Remove another comment

(I won't implement that check because I consider that an edge case, which
isn't worth a lot of effort.)

pabzm commented:
I fixed some more styling issues and handling quirks and added buttons to move a folder up and down the list (for people that can't or don't want to use drag-and-drop).

From my point of view this is now ripe for a proper review and hopefully merging, too.

pabzm commented:
> It does not seem to be finished. It has a lot of issues at this point.

@alecpl Could you let me know which those are?

pabzm committed:
Add semicolons to conform to the coding style

alecpl commented:
- That extra spacing (row height) is not good.
- Dragging a folder that wasn't selected creates bad look, i.e. the dragged object is transparent. But if it is selected it's kinda hard to get what's going on behind it (where the drop areas appear). Also, sometimes it's hard do match the dragging object with the drop area (especially when dealing with folder hierarchies).
- Dropping the folder scrolls the list up.
- Dropping into the initial place should probably not send a request (and not show "Successfully saved").
- I notice some jumping effect (height of rows changes) when hovering over a sub-tree (folder with sub-folders).
- The Up/Down buttons in folder info are kinda useless if you can't see the list on the same screen, i.e. on a phone.
- Mouse-up event outside the list does not stop the drag action, but should.

pabzm committed:
fixup! A little refactoring (folder_name2id())

pabzm committed:
Stop dragging of sortable items if mouseup happens over an iframe

pabzm committed:
Save sortable list only if it actually changed

pabzm committed:
Fix backgrounds and margins of selected subscriptionlist element in dark mode

alecpl commented:
> 1. I think the extra spacing is good, because it allows to more easily move folders between other folders. Without the padding there's only a few pixels of height to target, which are hard to find.

I think it's really bad. It's inconsistent with the rest. And drag effect is not nice either. I think this makes UX worse.

pabzm commented:
Thank you for the explanation! I'll have a look if I can integrate your suggestions.

pabzm committed:
Fix left indention of placeholder when sorting

pabzm committed:
Reduce line height of sortable folders list

pabzm committed:
Improve code to prevent sorting protected folders

pabzm committed:
Fix left indentation of items that were newly moved into a previously empty
list

pabzm committed:
Move sortable folder list code to treelist.js and adapt to previous UX
